win32/src/win32evtlog.i

```diff
@@ -511,6 +512,113 @@ static PyObject *PyEvtOpenChannelEnum(PyObject *self, PyObject *args, PyObject *
+// Caller must free the returned value
+static LPWSTR FormatMessageInternal(EVT_HANDLE metadata, EVT_HANDLE event, DWORD flags, DWORD resourceId)
+{
+	LPWSTR buf = NULL;
+	DWORD allocated_size = 0;
+	DWORD returned_size = 0;
+	DWORD status = 0;
+	DWORD err = 0;
+
+	BOOL bsuccess = 0;
+	Py_BEGIN_ALLOW_THREADS
+	// Get the size of the buffer
+	bsuccess = EvtFormatMessage(metadata, event, resourceId, 0, NULL, flags, allocated_size, buf, &returned_size);
+	Py_END_ALLOW_THREADS
+
+	err = GetLastError();
+	// The above call should always return ERROR_INSUFFICIENT_BUFFER
+	if (!bsuccess && err == ERROR_INSUFFICIENT_BUFFER) {
+		allocated_size = returned_size;
+		if (flags == EvtFormatMessageKeyword) {
+			allocated_size += 1; // +1 to double terminate the keyword list
+		}
+
+		allocated_size *= sizeof(WCHAR);
+		buf = (WCHAR *)malloc(allocated_size);
+		if (buf == NULL) {
+			PyErr_NoMemory();
+			return NULL;
+		}
+
+		Py_BEGIN_ALLOW_THREADS
+		bsuccess = EvtFormatMessage(metadata, event, resourceId, 0, NULL, flags, allocated_size, buf, &returned_size);
+		Py_END_ALLOW_THREADS
+
+		if (bsuccess) {
+			if (flags == EvtFormatMessageKeyword) {
+				buf[returned_size] = L'\0';
+			}
+		} else {
+			char buf[2048];
+			sprintf(buf, "EvtFormatMessage: allocated %d, need buffer of size %d", allocated_size, returned_size);
+			PyWin_SetAPIError(buf, err);
+		}
+	} else {
+		PyWin_SetAPIError("EvtFormatMessage" );
+	}
+
+	return buf;
+}
+
+// @pyswig str,list|EvtFormatMessage|Formats a message string.
+// @rdesc Returns a formatted message string, or a list of strings if Flags=EvtFormatMessageKeyword
+// @comm Accepts keyword args
+static PyObject *PyEvtFormatMessage(PyObject *self, PyObject *args, PyObject *kwargs)
+{
+	static char *keywords[] = {"Metadata", "Event", "Flags", "ResourceId", NULL};
+	EVT_HANDLE metadata_handle = NULL;
+	EVT_HANDLE event_handle = NULL;
+	DWORD flags = 0;
+	DWORD resourceId = 0;
+	if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O&O&k|k:EvtFormatMessage", keywords,
+		PyWinObject_AsHANDLE, &metadata_handle,	// @pyparm <o PyEVT_HANDLE>|Metadata||Handle to provider metadata returned by <om win32evtlog.EvtOpenPublisherMetadata>
+		PyWinObject_AsHANDLE, &event_handle,	// @pyparm <o PyEVT_HANDLE>|Event||Handle to an event
+		&flags,	// @pyparm int|Flags||Type of message to format. EvtFormatMessageEvent or EvtFormatMessageLevel or EvtFormatMessageTask or EvtFormatMessageOpcode or EvtFormatMessageKeyword or EvtFormatMessageChannel or EvtFormatMessageProvider or EvtFormatMessageId or EvtFormatMessageXml.  If set to EvtFormatMessageId, callers should also set the 'ResourceId' parameter
+        &resourceId))  // @pyparm int|ResourceId|0|The resource identifier of a message string returned by <om win32evtlog.EvtGetPublisherMetadataProperty>.  Only set this if flags = EvtFormatMessageId.
+		return NULL;
+
+	PyObject *ret = NULL;
+
+	LPWSTR buf = FormatMessageInternal(metadata_handle, event_handle, flags, resourceId);
+
+	if (buf) {
```

Comment:
Is there any reason to not have FormatMessageInternal return a PyObject? It would keep all the `flags` checks and memory management in that function.